From ec3566b75dd363127c7ba57a8e6b7f289078ce3f Mon Sep 17 00:00:00 2001
From: Guilherme Costa <guilherme.costa@windriver.com>
Date: Fri, 13 Dec 2024 11:57:25 +0000
Subject: [PATCH] csm-replication: add cluster secret template

Adding template that uses a job to that create the cluster
secret based on the Service Account data.

Signed-off-by: Guilherme Costa <guilherme.costa@windriver.com>
---
 .../templates/secret_setup.yaml               | 444 ++++++++++++++++++
 1 file changed, 444 insertions(+)
 create mode 100644 charts/csm-replication/templates/secret_setup.yaml

diff --git a/charts/csm-replication/templates/secret_setup.yaml b/charts/csm-replication/templates/secret_setup.yaml
new file mode 100644
index 0000000..98cc21c
--- /dev/null
+++ b/charts/csm-replication/templates/secret_setup.yaml
@@ -0,0 +1,444 @@
+{{/*
+#
+# Copyright (c) 2024 Wind River Systems, Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+*/}}
+{{ if .Values.config.targets }}
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: secret-creation-sa-role
+  namespace: {{ .Release.Namespace }}
+rules:
+- apiGroups: [""]
+  resources: ["namespaces"]
+  verbs: ["get", "list", "watch"]
+- apiGroups: [""]
+  resources: ["secrets"]
+  verbs: ["get", "create", "list", "update", "delete"]
+- apiGroups: [""]
+  resources: ["serviceaccounts"]
+  verbs: ["get", "list", "watch"]
+- apiGroups: ["batch"]
+  resources: ["jobs"]
+  verbs: ["get", "list", "create", "delete"]
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name:  secret-creation-sa
+  namespace: {{ .Release.Namespace }}
+imagePullSecrets:
+  - name: default-registry-key
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: secret-creation-sa-rb
+  namespace: {{ .Release.Namespace }}
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: secret-creation-sa-role
+subjects:
+- kind: ServiceAccount
+  name:  secret-creation-sa
+  namespace: {{ .Release.Namespace }}
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: secret-sa-creation-configmap
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "meta.helm.sh/release-name": {{ .Release.Name }}
+    "meta.helm.sh/release-namespace": {{ .Release.Namespace }}
+data:
+  gen_secret_from_sa.sh: |
+    #!/bin/sh
+
+    SECRET_YAML=/wd/secret_sa.yaml
+
+    echo "Removing secret ${SECRET_NAME} (if it already exists)"
+    kubectl delete secrets -n ${NAMESPACE} ${SECRET_NAME}
+
+    echo "Running gen_kubeconfig.sh"
+    bash -x /tmp/gen_kubeconfig.sh -s ${SERVICE_ACCOUNT} -n ${NAMESPACE} -o ${SECRET_YAML}
+
+    echo "Creating secret: ${SECRET_NAME}"
+    kubectl create secret generic ${SECRET_NAME} --from-file=data=${SECRET_YAML} -n ${NAMESPACE}
+
+  config-placeholder: |
+    apiVersion: v1
+    kind: Config
+    users:
+    - name: svcs-acct
+      user:
+        token: ${gntoken}
+    clusters:
+    - cluster:
+        certificate-authority-data: ${gncert}
+        server: ${gnserver}
+      name: self-hosted-cluster
+    contexts:
+    - context:
+        cluster: self-hosted-cluster
+        user: svcs-acct
+      name: svcs-acct-context
+
+  gen_kubeconfig.sh: |
+    #!/bin/sh
+    ######################################################################
+    # Copyright © 2021-2023 Dell Inc. or its subsidiaries. All Rights Reserved.
+    #
+    # Licensed under the Apache License, Version 2.0 (the "License");
+    # you may not use this file except in compliance with the License.
+    # You may obtain a copy of the License at
+    #      http://www.apache.org/licenses/LICENSE-2.0
+    # Unless required by applicable law or agreed to in writing, software
+    # distributed under the License is distributed on an "AS IS" BASIS,
+    # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    # See the License for the specific language governing permissions and
+    # limitations under the License.
+    ######################################################################
+
+    # Determine the command to use.
+    if [ -x "$(command -v kubectl)" ]; then
+      cmd='kubectl'
+    elif [ -x "$(command -v oc)" ]; then
+      cmd='oc'
+    else
+      echo "Neither kubectl nor oc is installed. Verify setup."
+      exit 1
+    fi
+
+    #######################################
+    # Prints formatted command's action
+    # Arguments:
+    #   $1 - custom message
+    #######################################
+    function t_print_action() {
+      printf "%-50s" "$1"
+    }
+
+    #######################################
+    # Prints formatted command's status
+    # Globals:
+    #   green - predefined color variable
+    #   nc - predefined color variable
+    # Arguments:
+    #   $1 - custom message
+    #######################################
+    function t_print_status() {
+      if [ -z "${1}" ]; then
+        printf "%-55s\n" "${green}[  OK  ]${nc}"
+      else
+        printf "%-55s\n" "$1"
+      fi
+    }
+
+    #######################################
+    # Prints error statement
+    # Globals:
+    #   nc - predefined color variable
+    #   red - predefined color variable
+    # Arguments:
+    #  None
+    #######################################
+    function finish_error() {
+      echo -e "${red}[ FAIL ]${nc}"
+      exit
+    }
+
+    #######################################
+    # Generates kubeconfig for service account in k8s.
+    # Globals:
+    #   gncert - CA's certificate
+    #   gnsecret - SA's secret
+    #   gnserver - server's URL
+    #   gntoken - SA's token
+    #   ns - (optional) SA'a namespace
+    #   output - output file location
+    #   o - output file location (CLI arg)
+    #   s - service account name (CLI arg)
+    # Arguments:
+    #  None
+    #######################################
+    function service_branch() {
+      if [ -z "${o}" ]; then
+        output="./configs/config-$s"
+      fi
+
+      t_print_action "Compiling kubeconfig for \"$s\"" && t_print_status
+      echo "Namespace: $ns, cmd: "
+      t_print_action "Extracting secret"
+      gnsecret=$($cmd -n $ns describe sa $s 2>/dev/null | grep Mountable | awk '{print $3}') && t_print_status || finish_error
+      t_print_action "Extracting token"
+      gntoken=$($cmd -n $ns describe secret $gnsecret | grep token: | awk '{print $2}') && t_print_status || finish_error
+      t_print_action "Extracting CA"
+      gncert=$($cmd config view --flatten --minify | grep certificate-authority-data: | awk '{print $2}') && t_print_status || finish_error
+      t_print_action "Extracting server URL"
+      gnserver=$($cmd config view --flatten --minify | grep server: | awk '{print $2}') && t_print_status || finish_error
+
+      export gnsecret
+      export gntoken
+      export gncert
+      export gnserver
+
+      folder=$(dirname "$0")
+      envsubst >"$output" < "$folder"/config-placeholder
+      echo "DONE. Kubeconfig location: $output"
+
+    }
+
+    #######################################
+    # Generates kubeconfig for "normal" user in k8s.
+    # Globals:
+    #   base64_csr - base64 encoded CSR
+    #   certname - CN to use for kubeconfig creation
+    #   output - output file location
+    #   c - Certificate Signing Request
+    #   k - user's private key
+    #   o - output file location
+    #   u - user's username
+    # Arguments:
+    #  None
+    #######################################
+    function user_branch() {
+      if [ -z "${c}" ] || [ -z "${k}" ]; then
+        usage
+      fi
+
+      if [ -z "${o}" ]; then
+        output="./configs/config-$u"
+      fi
+
+      t_print_action "Encoding and applying CSR.. "
+      export base64_csr=$(cat $c | base64 | tr -d '\n')
+      export certname=$u
+      cat csr.yaml | envsubst | $cmd apply -f - &>/dev/null && t_print_status || finish_error
+
+      t_print_action "Approving CSR.. "
+      $cmd certificate approve $u &>/dev/null && t_print_status || finish_error
+
+      t_print_action "CSR Status: "
+      t_print_status "$($cmd get csr | grep $u | awk '{print $5}')"
+
+      t_print_action "Saving cert.. "
+      $cmd get csr $u -o jsonpath='{.status.certificate}' | base64 --decode >./tmp/$u.crt && t_print_status || finish_error
+
+      t_print_action "Using name \"$u\".. "
+      t_print_status
+
+      t_print_action "Saving temp CA cert.. "
+      $cmd config view -o jsonpath='{.clusters[0].cluster.certificate-authority-data}' --raw | base64 --decode - >./tmp/kubernetes-ca.crt && t_print_status || finish_error
+
+      t_print_action "Creating kubeconfig for \"$u\".. "
+      $cmd config set-cluster $($cmd config view -o jsonpath='{.clusters[0].name}') --server=$($cmd config view -o jsonpath='{.clusters[0].cluster.server}') --certificate-authority=./tmp/kubernetes-ca.crt --kubeconfig=$output --embed-certs &>/dev/null && t_print_status || finish_error
+
+      t_print_action "Setting credentials.. "
+      $cmd config set-credentials $u --client-certificate=./tmp/$u.crt --client-key=$k --embed-certs --kubeconfig=$output &>/dev/null && t_print_status || finish_error
+
+      t_print_action "Setting context.. "
+      $cmd config set-context $u --cluster=$($cmd config view -o jsonpath='{.clusters[0].name}') --namespace=default --user=$u --kubeconfig=$output &>/dev/null && t_print_status || finish_error
+
+      t_print_action "Applying context.. "
+      $cmd config use-context $u --kubeconfig=$output &>/dev/null && t_print_status || finish_error
+      echo "DONE. Kubeconfig location: $output"
+    }
+
+    #######################################
+    # Prints usage information and exits
+    #######################################
+    function usage() {
+      echo "Usage: $0 (( -u <CN user> -c <CSR> -k <key> || -s <serviceAccount> [-n <namespace>] )) && [-o <path/to/outputfile>]" 1>&2
+      exit 1
+    }
+
+    #######################################
+    # Main entrypoint
+    # Arguments:
+    #  None
+    #######################################
+    function main() {
+      set -o pipefail
+
+      ns=default
+
+      red=$(tput setaf 1)
+
+      green=$(tput setaf 2)
+
+      nc=$(tput sgr0)
+
+      mkdir -p ./tmp
+
+      mkdir -p ./configs
+
+      while getopts ":u:s:o:n:c:k:" i; do
+        case "${i}" in
+          u)
+            u=${OPTARG}
+            ;;
+          c)
+            c=${OPTARG}
+            ;;
+          k)
+            k=${OPTARG}
+            ;;
+          s)
+            s=${OPTARG}
+            ;;
+          n)
+            n=${OPTARG}
+            ns=$n
+            ;;
+          o)
+            o=${OPTARG}
+            output="$o"
+            ;;
+          *)
+            usage
+            ;;
+        esac
+      done
+
+      shift $((OPTIND - 1))
+
+      if [ -z "${u}" ] && [ -z "${s}" ]; then
+        usage
+      fi
+
+      if [ -z "${u}" ]; then
+        service_branch
+      else
+        user_branch
+      fi
+
+      rm -rf ./tmp
+
+    }
+
+    main "$@"
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: "secret-sa-creation-job"
+  namespace: {{ .Release.Namespace }}
+  labels:
+    control-plane: controller-manager
+    app: "{{ $.Chart.Name }}"
+    app.starlingx.io/component: {{ ternary "application" "platform" .Values.labels.isApplication }}
+    chart: "{{ $.Chart.Name }}-{{ $.Chart.AppVersion }}"
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+  annotations:
+    "meta.helm.sh/release-name": {{ $.Release.Name }}
+    "meta.helm.sh/release-namespace": {{ $.Release.Namespace }}
+    "helm.sh/hook-weight": "1"
+    "helm.sh/hook": "post-install,pre-upgrade,pre-rollback"
+    "helm.sh/hook-delete-policy": "before-hook-creation,hook-succeeded"
+spec:
+  backoffLimit: 5
+  activeDeadlineSeconds: 100
+  template:
+    spec:
+      serviceAccountName: secret-creation-sa
+      securityContext:
+          fsGroup: 345
+      volumes:
+      - name: secret-sa-creation-configmap-volume
+        configMap:
+          name: secret-sa-creation-configmap
+          defaultMode: 0555
+      - name: workdir-volume
+        emptyDir: {}
+      - name: kube-config
+        hostPath:
+          path: /etc/kubernetes/admin.conf
+          type: File
+      containers:
+        - name: create-cluster-local-sa-secret
+          image: {{ $.Values.kubectl.image }}
+          imagePullPolicy: IfNotPresent
+          workingDir: /wd
+          command: ["/bin/sh", "/tmp/gen_secret_from_sa.sh" ]
+          env:
+            - name: NAMESPACE
+              value: {{ $.Release.Namespace }}
+            - name: SERVICE_ACCOUNT
+              value: dell-replication-controller
+            - name: SECRET_NAME
+              value: {{ printf "cluster-local-sa-secret-%s" $.Values.config.clusterId | trunc 63 | trimSuffix "-" }}
+            - name: KUBECONFIG
+              value: /etc/kubernetes/admin.conf
+          volumeMounts:
+          - name: secret-sa-creation-configmap-volume
+            mountPath: /tmp
+            readOnly: false
+          - name: workdir-volume
+            mountPath: /wd
+            readOnly: false
+          - name: kube-config
+            mountPath: /etc/kubernetes/admin.conf
+            readOnly: true
+      restartPolicy: OnFailure
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: sa-secret-cleanup
+  namespace: {{ .Release.Namespace }}
+data:
+  sa_secret_clean_up.sh: |-
+    #!/bin/sh
+
+    kubectl delete -n ${NAMESPACE} secrets ${SECRET_NAME}
+
+    echo "Secret cleanup: ${SECRET_NAME}"
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: "secret-sa-delete-job"
+  namespace: {{ $.Release.Namespace }}
+  labels:
+    control-plane: controller-manager
+    app: "{{ $.Chart.Name }}"
+    app.starlingx.io/component: {{ ternary "application" "platform" .Values.labels.isApplication }}
+    chart: "{{ $.Chart.Name }}-{{ $.Chart.AppVersion }}"
+    release: {{ $.Release.Name }}
+    heritage: {{ $.Release.Service }}
+  annotations:
+    "meta.helm.sh/release-name": {{ $.Release.Name }}
+    "meta.helm.sh/release-namespace": {{ $.Release.Namespace }}
+    "helm.sh/hook": "pre-delete"
+    "helm.sh/hook-delete-policy": "before-hook-creation,hook-succeeded,hook-failed"
+spec:
+  backoffLimit: 5
+  activeDeadlineSeconds: 100
+  template:
+    spec:
+      serviceAccountName: secret-creation-sa
+      volumes:
+        - name: sa-secret-cleanup
+          configMap:
+            name: sa-secret-cleanup
+      containers:
+        - name: "secret-sa-cleanup"
+          image: {{ $.Values.kubectl.image }}
+          imagePullPolicy: IfNotPresent
+          command: ["/bin/sh", "/tmp/mount/sa_secret_clean_up.sh"]
+          env:
+            - name: NAMESPACE
+              value: {{ $.Release.Namespace }}
+            - name: SECRET_NAME
+              value: {{ printf "cluster-local-sa-secret-%s" $.Values.config.clusterId | trunc 63 | trimSuffix "-" }}
+          volumeMounts:
+            - name: sa-secret-cleanup
+              mountPath: /tmp/mount
+      restartPolicy: Never
+{{ end }}
-- 
2.34.1
